.github/workflows/test_precommit.yml

```diff
@@ -108,7 +109,7 @@ jobs:
     - uses: actions/setup-python@v4
       with:
         python-version: 3.9
-        #cache: pip
+        cache: pip
```

Comment:
Just in case, I don't think we need caching at all. Enabling the cache just downloads an archive from GitHub instead of PyPI. Moreover as far as I understand, modern pip downloads packages asynchronously all at the same time which may even outperform archive download.

Reply:
Haven't heard pip is parallel, I found a comparison with other python package installers here: https://pythonspeed.com/articles/faster-pip-installs/
pip logs also make an impression the packages are downloaded installed sequentially.
pip is faster when cache is not empty, according to the paper. I'd expect, GHA could fetch it's own cache faster than pip downloads packages from some 3rd party mirrors

Reply:
https://github.com/pypa/pip/issues/825

Reply:
Ok. Up to you